### PR TITLE
🌱 Add high level description to e2e tests

### PR DESCRIPTION
**Pull request by mboukhalfa**
Add human readable description of e2e tests

### CODE REVIEW

File: test/e2e/integration_test.go [hunk 1]
```diff
@@ -9,6 +9,17 @@ import (
...
 
+// Integration tests in capm3 focus on validating the seamless integration between different components of the capm3 project,
+// including capm3, IPAM, CAPI, BMO, and Ironic. These tests ensure that these components work together cohesively to provision a
+// workload cluster and perform pivoting operations between the bootstrap cluster and the target cluster.
+// The primary goal is to detect any compatibility issues or conflicts that may arise during integration.
+
+// By executing the integration tests, capm3 verifies that:
+
+// - The capm3 controller effectively interacts with the IPAM, CAPI, BMO, and Ironic components.
+// - The provisioning of a workload cluster proceeds smoothly, creating bmhs, inspecting and provision them.
```
Comment (huutomerkki) [inline test/e2e/integration_test.go:20]: This line doesn't really make sense as it is.

File: test/e2e/integration_test.go [hunk 2]
```diff
@@ -9,6 +9,17 @@ import (
...
+// - The pivoting functionality enables the seamless transition of resources and control components from the bootstrap cluster to the target cluster and vice versa.
+// - Deprovision then cluster and wait for the bmh to get deprovisoned.
```
Comment (huutomerkki) [inline test/e2e/integration_test.go:22]: Neither does this. Maybe something like "Deprovisioning the cluster and bmhs happens smoothly."

File: test/e2e/live_iso_test.go [hunk 3]
```diff
@@ -18,6 +18,8 @@ var _ = Describe("When testing live iso [live-iso] [features]", func() {
 	liveIsoTest()
 })
 
+// Live iso tests provision live-iso image on host
```
Comment (huutomerkki) [inline test/e2e/live_iso_test.go:21]: I think the scope of the test could be clarified a bit more here.

File: test/e2e/pivoting.go [hunk 4]
```diff
@@ -50,6 +50,7 @@ type PivotingInput struct {
 	ClusterctlConfigPath  string
 }
 
+// Pivoting implements a test that verifies moving management resources (CRs, BMO, Ironic) to a target cluster after initializing it with Provider components.
```
Comment (huutomerkki) [inline test/e2e/pivoting.go:53]: Nitpicking a bit, but verifies what exactly?

File: test/e2e/upgrade_baremetal_operator.go [hunk 5]
```diff
@@ -17,6 +17,7 @@ type upgradeBMOInput struct {
 	SpecName          string
 }
 
+// upgradeIronic upgrade BMO image to the latest.
```
Comment (huutomerkki) [inline test/e2e/upgrade_baremetal_operator.go:20]: A typo here, should likely be upgradeBMO

**Review by huutomerkki — COMMENTED** [review]
The bigger, detailed comments were good and informative.

File: test/e2e/metal3remediation.go
Comment (mboukhalfa) [inline test/e2e/metal3remediation.go:1]: @slintes Can you please take a look if I well covered the test in the description
 also I am planning to add full health-check to e2e https://github.com/metal3-io/cluster-api-provider-metal3/issues/1060 and your suggestion about keep this test as it is (explicitly test node deletion ) or refactoring it (to be implicit in the health-check) would be helpful

File: test/e2e/integration_test.go (hunk 1 above)
Comment (mboukhalfa) [reply test/e2e/integration_test.go:20]: Trying to show that the integration test verifies that we can create workload cluster using metal3 stack, starting by creating the BMH and then inspecting them and at the end provision the image image ?

File: test/e2e/integration_test.go (hunk 2 above)
Comment (mboukhalfa) [reply test/e2e/integration_test.go:22]: done

File: test/e2e/pivoting.go (hunk 4 above)
Comment (mboukhalfa) [reply test/e2e/pivoting.go:53]: it verifies "moving management resources" the movement of crs to target cluster

File: test/e2e/upgrade_baremetal_operator.go (hunk 5 above)
Comment (mboukhalfa) [reply test/e2e/upgrade_baremetal_operator.go:20]: copy paste 😄

File: test/e2e/integration_test.go (hunk 1 above)
Comment (huutomerkki) [reply test/e2e/integration_test.go:20]: More about the wording then, maybe something like
"The provisioning of a workload cluster proceeds smoothly, and that bmhs are created, inspected and provisioned as expected."

File: test/e2e/pivoting.go (hunk 4 above)
Comment (huutomerkki) [reply test/e2e/pivoting.go:53]: "Pivoting implements a test that verifies successful moving of management resources (CRs, BMO, Ironic) to a target cluster after initializing it with Provider components." or something along those lines is what I was after.

File: test/e2e/integration_test.go (hunk 1 above)
Comment (mboukhalfa) [reply test/e2e/integration_test.go:20]: done

File: test/e2e/pivoting.go (hunk 4 above)
Comment (mboukhalfa) [reply test/e2e/pivoting.go:53]: thanks

File: test/e2e/metal3remediation.go [hunk 6]
```diff
@@ -27,6 +27,34 @@ type Metal3RemediationInput struct {
...
+ * This test evaluates the reboot remediation strategy with an enhancement of node deletion in the CAPM3 (Cluster API Provider for Metal3) Remediation Controller.
+ *
+ * Tested Feature:
+ * - Delete Node in Reboot Remediation
+ *
+ * Workflow:
+ * 1. Retrieve the Metal3Machines associated with the worker nodes in the cluster.
+ * 2. Identify the target worker machine node its associated BMH object corresponding to the Metal3Machine.
+ * 3. Create a Metal3Remediation resource, specifying the remediation strategy as "Reboot" with a retry limit and timeout.
+ * 4. Wait for the VMs (Virtual Machines) associated with target BMH to power off.
```
Comment (slintes) [inline test/e2e/metal3remediation.go:45]: nit: its one VM only, so singular `VM` and `Virtual Machine` would be a slightly better fit. Same below for power on

File: test/e2e/metal3remediation.go
Comment (slintes) [reply test/e2e/metal3remediation.go:1]: - overall lgtm nut I left a comment (very small nit only)
- I don't have any context (and no time atm to get it, sorry) about 1060, but in general node deletion is an important part of the remediation process and should be tested explicitly somewhere IMHO.

File: test/e2e/metal3remediation.go (hunk 6 above)
Comment (mboukhalfa) [reply test/e2e/metal3remediation.go]: Fixed thanks

**Review by kashifest — APPROVED** [review]

**Review by Sunnatillo — APPROVED** [review]
/lgtm
Excellent work @mboukhalfa

File: test/e2e/node_reuse.go [hunk 7]
```diff
@@ -32,6 +32,7 @@ type NodeReuseInput struct {
 	Namespace         string
 }
 
+// NodeReuse verifies the feature of reusing the same node after upgrading kcp nodes.
```
Comment (furkatgofurov7) [inline test/e2e/node_reuse.go:35]:
```suggestion
// NodeReuse verifies the feature of reusing the same node after upgrading kcp/md nodes.
```

File: test/e2e/live_iso_test.go [hunk 8]
```diff
@@ -14,10 +14,28 @@ import (
...
+ * 	Serial logs are read to verify that the node was booted from the live ISO image.
+ * 	The test is considered passed.
+ */
+
 var _ = Describe("When testing live iso [live-iso] [features]", func() {
 	liveIsoTest()
 })
 
+// Live iso tests provision live-iso image on host
+// it lists all the bmh and selects the one supporting redfish-virtualmedia for provisioning the live image.
```
Comment (furkatgofurov7) [inline test/e2e/live_iso_test.go:38]: is this needed, I see it covered above?

File: test/e2e/pivoting_based_feature_test.go [hunk 9]
```diff
@@ -26,6 +26,57 @@ var (
...
+ * Certificate Rotation:
+ * This test ensures that certificate rotation in the Ironic pod is functioning correctly.
+ * by forcing certificate regeneration and verifying container restarts
+ * - It starts by checking if the Ironic pod is running. It retrieves the Ironic deployment and waits for the pod to be in the "Running" state.
+ * - The test forces cert-manager to regenerate the certificates by deleting the relevant secrets.
+ * - It then waits for the containers in the Ironic pod to be restarted. It checks if each container exists and compares the restart count with the previously recorded values.
+ * - If all containers are restarted successfully, the test passes.
+ *
+ * Node Reuse:
+ * This test verifies the feature of reusing the same node after upgrading Kubernetes control plane (KCP) nodes.
```
Comment (furkatgofurov7) [inline test/e2e/pivoting_based_feature_test.go:58]:
```suggestion
 * This test verifies the feature of reusing the same node after upgrading Kubernetes version in KubeadmControlPlane (KCP) and MachineDeployment (MD) nodes.
```

File: test/e2e/metal3remediation.go [hunk 10]
```diff
@@ -27,6 +27,34 @@ type Metal3RemediationInput struct {
 	Namespace             string
 }
 
+/*
+ * Metal3 Remediation Test
+ *
+ * This test evaluates node deletion in reboot remediation feature added to CAPM3 Remediation Controller.
+ * issue #392: Reboot remediation is incomplete
+ * PR #668: Fix reboot remediation by adding node deleti
+ * This test evaluates the reboot remediation strategy with an enhancement of node deletion in the CAPM3 (Cluster API Provider for Metal3) Remediation Controller.
```
Comment (furkatgofurov7) [inline test/e2e/metal3remediation.go:36]: These looks incomplete and not needed, perhaps we can remove all of it?

**Review by furkatgofurov7 — COMMENTED** [review]
This is great and in detail step-by-step skeleton of each implemented feature that gives nice high level view, thanks @mboukhalfa!. Few nits / suggestions we could incorporate:

**Comment by furkatgofurov7** [comment]
/retitle 🌱 Add high level description to e2e tests

**Comment by furkatgofurov7** [comment]
/hold 

to see if we can address the comments with the same PR (feel free to unhold if you want to do them in a follow-up)

File: test/e2e/live_iso_test.go (hunk 8 above)
Comment (mboukhalfa) [reply test/e2e/live_iso_test.go:38]: I would like to keep this docs for the function while above we are documenting the test file or the describe block

**Review by furkatgofurov7 — COMMENTED** [review]
Thanks @mboukhalfa, looks good. 
I also noticed, few other e2e tests we have (specifically: IP reuse, ~inspection~) is not covered in this PR while it could be beneficial to add the descriptions for those tests as well. That being said, it is not a blocker and can be done in a follow-up if needed.

**Comment by mboukhalfa** [comment]
> Thanks @mboukhalfa, looks good. I also noticed, few other e2e tests we have (specifically: IP reuse, ~inspection~) is not covered in this PR while it could be beneficial to add the descriptions for those tests as well. That being said, it is not a blocker and can be done in a follow-up if needed.

Yeah let's do it in follow up

File: test/e2e/inspection.go [hunk 11]
```diff
@@ -24,6 +24,7 @@ type InspectionInput struct {
 	SpecName              string
 }
 
+// Inspection test request inspection on all the available BMH using annotation.
```
Comment (furkatgofurov7) [inline test/e2e/inspection.go:27]:
```suggestion
// Inspection test requests inspection on all the available BMH using annotation.
```

**Review by furkatgofurov7 — APPROVED** [review]
/approve

**Pull request merged**